src/lib/core/components/visualizations/InputVariables.tsx

```diff
@@ -137,10 +135,37 @@ export function InputVariables(props: Props) {
   > = useMemo(
     () =>
       inputs.reduce((map, input) => {
-        const disabledVariables = excludedVariables(
-          entities[0],
-          flattenedConstraints && flattenedConstraints[input.name]
-        );
+        // For each input (ex. xAxisVariable), determine its constraints based on which patterns any other selected variables match.
+        const filteredConstraints =
+          constraints &&
+          filterConstraints(
+            selectedVariables,
+            entities,
+            constraints,
+            input.name
+          );
+
+        // Use the input-specific filtered constraints to create an array of disabled variables.
+        const disabledVariables = filteredConstraints
+          ? filteredConstraints
+              .map((constraint) => {
+                return excludedVariables(
+                  entities[0],
+                  constraint && constraint[input.name]
+                );
+              })
+              // Keep only those variables (objects with varIds and entityIds) that should be disabled based on *all* constraints
+              .reduce((disabledVarArray1, disabledVarArray2) =>
+                disabledVarArray1.filter((disabledVar1) =>
+                  disabledVarArray2.some(
+                    (disabledVar2) =>
+                      disabledVar1.variableId === disabledVar2.variableId &&
+                      disabledVar1.entityId === disabledVar2.entityId
+                  )
+                )
+              )
+          : [{ entityId: '', variableId: '' }]; // Typescript help please!!! I tried [{}] as DataElementConstraintRecord but no luck :/
```

Comment:
I would take a different approach. Instead of modifying this logic, I would modify `excludedVariables` so that it takes an array of `DataElementConstraint`:

```typescript
export function excludedVariables(
  rootEntity: StudyEntity,
  constraints?: DataElementConstraint[]
): VariableDescriptor[]
```

I think it lead to more flexible code in the long run.